chill-hadoop/src/main/java/com/twitter/chill/hadoop/KryoSerialization.java

```diff
@@ -38,6 +38,27 @@
     // can't be final because we need to set them in setConf (for Configured)
     KryoPool kryoPool;
     Kryo testKryo;
+
+    private static KryoPool cachedPool = null;
+    private static Kryo cachedTestKryo = null;
+    private static KryoInstantiator cachedKryoInst = null;
+
+    /**
+     * Hadoop will re-initialize the KryoSerialization on every spill
+     * This gets very expensive if you output a lot from a mapper to initialize the chill/kryo stack
+     * The KryoInstantiator's already do some caching, and figuring out if its safe to cache,
+     * so here we piggy back on that to avoid generating new kryo's or kryo pools
+     */
+    public static synchronized void resetOrUpdateFromCache(KryoSerialization instance, KryoInstantiator kryoInst){
+      if(kryoInst != cachedKryoInst) {
+        cachedTestKryo = kryoInst.newKryo();
+        cachedPool = KryoPool.withByteArrayOutputStream(MAX_CACHED_KRYO, kryoInst);
+        cachedKryoInst = kryoInst;
+      }
+      instance.kryoPool = cachedPool;
+      instance.testKryo = cachedTestKryo;
```

Comment:
I'm worried about the thread-safety of testKryo. We could be sharing it across threads, and there is no guarantee it is thread-safe.

Could we either borrow from the KryoPool to test or use a lock to guard access?